### PR TITLE
Wrap scribble computations in unchecked for Solidity >= 0.8.0

### PR DESCRIPTION
**Pull request by cd1m0**
After 0.8.0 Solidity arithmetic becomes 'checked' by default. In order to:

1. Have consistent behavior with Scribble pre-0.8.0
2. Ensure that spec computation does not risk throwing additional errors
3. Make it easier to write specs about functions containing unchecked computation

We have decided that scribble specs will have **unchecked** semantics. I.e. they may silently overflow without reverting.
To implement this, we wrap all scribble instrumentation-related computation in `unchecked {}` blocks.

### CODE REVIEW

**Comment by cd1m0** [comment]
CI timed out again. It finishes fine locally. Really need to do something about this :(

File: test/samples/if_assigned_complex.instrumented.sol [hunk 1]
```diff
@@ -60,83 +60,101 @@ contract Base {
 
     function Base_x_inline_initializer() internal {
         vars1 memory _v;
-        _v.old_0 = x;
-        if (!(x >= _v.old_0)) {
-            emit AssertionFailed("0: ");
-            assert(false);
+        unchecked {
+            _v.old_0 = x;
+        }
```
Comment (blitz-1306) [inline test/samples/if_assigned_complex.instrumented.sol:65]: It looks a bit strange, because I do not think if overflow/underflow may happen with such plain assignments. Is this a side effect from some changes in the code base?

File: test/samples/if_assigned_primitive.instrumented.sol [hunk 2]
```diff
@@ -55,91 +55,117 @@ contract Base {
...
+            if (!(y >= _v.old_1)) {
+                emit AssertionFailed("1: ");
+                assert(false);
+            }
         }
     }
 
     function Base_x_inc_postfix() internal returns (uint256 RET3) {
         RET3 = x;
         x++;
```
Comment (blitz-1306) [inline test/samples/if_assigned_primitive.instrumented.sol:127]: We can have overflow here for `++` and underflow for `--`, I guess. Should we also wrap unary operations (except `delete`) in `unchecked` block?

This is the first occurance, however there are more occurances in other snapshots. I didn't commented them all to not cause comment spamming.

**Review by blitz-1306 — COMMENTED** [review]
Good start! I think that we need to cover a bit more edge cases:
1. We should consider user-defined function (UDF) bodies. Their call sites may be wrapped by `unchecked`, as they would be a part of predicates. However their bodies may contain literally anything that would be executed in "checked" manner as `unchecked` state is not propagated to the called function.
2. We have unary operations not wrapped by `unchecked` in few samples. I suspect that this would break the initial promise.
3. The complicated part would be if UDFs or predicates are containing a call to some local function, that is executed in checked manner. Unsure if we then can handle it properly.

Let's take a bit more time to check that these assumptions are really applicable and handle the edge cases (if there any).

File: test/samples/if_assigned_complex.instrumented.sol (hunk 1 above)
Comment (cd1m0) [reply test/samples/if_assigned_complex.instrumented.sol:65]: This is a side effect of me just wrapping everything related to property computation in `unchecked` blocks, regardless of whether it contains arithmetic. Its just easier to write it that way.

File: test/samples/if_assigned_primitive.instrumented.sol (hunk 2 above)
Comment (cd1m0) [reply test/samples/if_assigned_primitive.instrumented.sol:127]: So this `++` corresponds to the **original** operation that we are interposing on. It is not wrapped in `unchecked` here because the original operation in `if_assigned_pritimitive.sol` was not wrapped in unchecked https://github.com/ConsenSys/scribble/blob/develop/test/samples/if_assigned_primitive.sol#L25. If you wrap line 25 in if_assigned_primitive.sol in an unchecked, then here `x++` will also be wrapped in unchecked.

Comment (cd1m0) [reply test/samples/if_assigned_primitive.instrumented.sol:127]: This however brings a new problem - we may need to emit 2 versions of wrappers - one version when the wrapper is inside of an unchecked, and another when its not inside an unchecked. Good catch!

**Comment by cd1m0** [comment]
> Good start! I think that we need to cover a bit more edge cases:
> 
>     1. We should consider user-defined function (UDF) bodies. Their call sites may be wrapped by `unchecked`, as they would be a part of predicates. However their bodies may contain literally anything that would be executed in "checked" manner as `unchecked` state is not propagated to the called function.
> 

Good call. I think UDF bodies should ALWAYS be inside unchecked. They are scribble code, and scribble arithmetic is by definition unchecked.

>     2. We have unary operations not wrapped by `unchecked` in few samples. I suspect that this would break the initial promise.
> 

The unary operations in the samples you are refering correspond to original operations in the solidity code, not scribble. So they should be inside an unchecked block, only IF the original expression was also in unchecked. But this also means that we may need to emit 2 versions of some wrappers. Will fix and add samples for this. Good catch!

>     3. The complicated part would be if UDFs or predicates are containing a call to some local function, that is executed in checked manner. Unsure if we then can handle it properly.
> 
I think this is ok. The "unchecked"-ness is only function-local. So if you have:

```
unchecked {
  uint x = foo(x + 1);
}

function foo(uint x) returns (uint) { return x + 1; }
```

Then the body of `foo()` is still executed as checked.

> 
> Let's take a bit more time to check that these assumptions are really applicable and handle the edge cases (if there any).

File: test/samples/if_assigned_primitive.instrumented.sol (hunk 2 above)
Comment (cd1m0) [reply test/samples/if_assigned_primitive.instrumented.sol:127]: Fixed in 0278d5c

**Review by blitz-1306 — APPROVED** [review]
